src/uct/ib/base/ib_device.c

```diff
@@ -238,22 +237,48 @@ uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
         goto out;
     }
 
+    ucs_assert(ret != UCS_KH_PUT_KEY_PRESENT);
+    entry       = &kh_value(&dev->async_events_hash, iter);
+    entry->cbq  = cbq;
+    entry->cb   = NULL;
+    entry->flag = 0;
+    status      = UCS_OK;
+
+out:
+    ucs_spin_unlock(&dev->async_event_lock);
+    return status;
+}
+
+ucs_status_t
+uct_ib_device_async_event_wait(uct_ib_device_t *dev,
+                               enum ibv_event_type event_type,
+                               uint32_t resource_id,
+                               ucs_callback_t cb, void *arg)
+{
+    uct_ib_async_event_val_t *entry;
+    uct_ib_async_event_t event;
+    ucs_status_t status;
+    khiter_t iter;
+
+    event.event_type  = event_type;
+    event.resource_id = resource_id;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter  = kh_get(uct_ib_async_event, &dev->async_events_hash, event);
+    ucs_assert(iter != kh_end(&dev->async_events_hash));
     entry = &kh_value(&dev->async_events_hash, iter);
-    if (ret == UCS_KH_PUT_KEY_PRESENT) {
-        if (entry->flag) {
-            status = UCS_OK;
-            goto out;
-        } else if (entry->cb) {
-            status = UCS_ERR_BUSY;
-            goto out;
-        }
+
+    if (entry->flag) {
+        status = UCS_OK;
+        goto out;
+    } else if (entry->cb != NULL) {
+        status = UCS_ERR_BUSY;
+        goto out;
```

Comment:
we can replace goto out by else:
if (flag){
 status=OK
} else if (cb!=NULL) {
 status=BUSY
} else {
 entry->cb=cb;
 ...
}